tests/cpp/test_multidevice_pipeline.cpp

```diff
@@ -132,9 +136,16 @@ TEST_F(PipelineTest, Pipeline) {
 
 //(backend type, first stage's mesh, second stage's mesh (if not null), is first
 // stage sharded?, is second
-// stage sharded?, do_reduction?, use_fusion_executor_cache?)
-using PipelineTestTwoStagesParams = std::
-    tuple<CommunicatorBackend, DeviceMesh, DeviceMesh, bool, bool, bool, bool>;
+// stage sharded?, do_reduction?, sharded dimension, use_fusion_executor_cache?)
+using PipelineTestTwoStagesParams = std::tuple<
+    CommunicatorBackend,
+    DeviceMesh,
+    DeviceMesh,
+    bool,
+    bool,
+    bool,
+    int,
+    bool>;
```

Comment:
This is a good suggestion, this test's params keep growing as we add new functionality. I'll add it now :D

Reply:
It totally make sense to define structs here, but as far as I can see this won't avoid things like `testing::Values(false)` in the test instantiation
But correct me if Im wrong !

Reply:
I was thinking about:
```
struct Config {
  knob1;
  knob2;
};

testing::Values({.knob1=..., .knob2=...}, {.knob1=..., .knob2=...})
```
or even
```
testing::ValuesIn(AFunctionThatGeneratesAVectorOfConfigs());
```

If `testing::Combine` is too handy, another way to enforce names is to use enum (e.g. `enum class ShardingType`) instead of booleans. After all, my main concern is that constants are floating around without names and types, so it's very easy to put `testing::Values(...)` in the wrong order.

Reply:
In my opinion Combine is very convenient here so I would keep it even if I admit that it is at the cost of readability when instantiating the test

Reply:
Another drawback of using structs or enum class is that we will need to overload the `<<` operator for each created type for readable prints, whereas `Combine` is a native gtest and verything gets printed nicely

Reply:
Googletest supports the struct parameters and `testing::combine`. It is not the cleanest design. I've pushed some code, but it doesn't get rid of `testing::Values(false)` since we are still using combine. 
To improve the readability we can add comments to the current test instantiations like:
```
testing::Values(true), // first_stage_sharded
testing::Values(true), // second_stage_sharded
testing::Values(false), // do_reduction
```

Reply:
We can revert back to using the original tuple parameters, since I don't think this improved the readability that much.